### PR TITLE
Add safer-cluster and safer-cluster-update-variant modules to autogen system

### PR DESCRIPTION
**Pull request by skinlayers**

### CODE REVIEW

**Comment by skinlayers** [comment]
@morgante I'm guessing the checks are failing because I need to also make changes to `./test/`? AFAICT, I can't actually see the Cloud Build output, so I don't know what the actual errors are.

**Comment by morgante** [comment]
@skinlayers You can also run tests locally, see the [contribution guide](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/CONTRIBUTING.md).

**Comment by skinlayers** [comment]
@morgante Ok. I'm blocked until I can get additional access in my company's GCP org perms/project folder structure to run `test/setup`. I currently cannot assign the "Project Creator" role to a service account.

**Comment by morgante** [comment]
@skinlayers If you yourself have Project Creator access, you could run `terraform apply` within `test/setup` to create a project.

The linters (`make docker_test_lint`) should also be testable without any GCP permissions.

File: autogen/safer-cluster/main.tf.tmpl [hunk 1]
```diff
@@ -107,7 +107,7 @@ module "gke" {
   cluster_resource_labels = var.cluster_resource_labels
 
   // We enable private endpoints to limit exposure.
-  enable_private_endpoint       = true
```
Comment (skinlayers) [inline autogen/safer-cluster/main.tf.tmpl]: This has the opposite effect of what its name implies. In order to follow the GKE hardening guides recommendation at https://cloud.google.com/kubernetes-engine/docs/how-to/hardening-your-cluster#restrict_network_access_to_the_control_plane_and_nodes ("Public endpoint access enabled, master authorized networks enabled"), this `enable_private_endpoint` needs to be set to `false`.


See also https://stackoverflow.com/a/58651940/6720110

File: autogen/safer-cluster/main.tf.tmpl [hunk 2]
```diff
@@ -151,4 +151,6 @@ module "gke" {
   authenticator_security_group = var.authenticator_security_group
 
   enable_shielded_nodes = var.enable_shielded_nodes
+
+  skip_provisioners = var.skip_provisioners
```
Comment (skinlayers) [inline autogen/safer-cluster/main.tf.tmpl]: We need the ability to disable the provisioners if we're running in Terraform Cloud/Terraform Enterprise as `jq` and `gcloud` cannot be easily installed.

File: autogen/safer-cluster/main.tf.tmpl [hunk 3]
```diff
@@ -130,9 +130,7 @@ module "gke" {
 
   // Define PodSecurityPolicies for differnet applications.
   // Example: https://kubernetes.io/docs/concepts/policy/pod-security-policy/#example
-  pod_security_policy_config = [{
-    "enabled" = true
-  }]
+  pod_security_policy_config = var.pod_security_policy_config
```
Comment (skinlayers) [inline autogen/safer-cluster/main.tf.tmpl]: This is a selfish change, but at least the default remains the same. The devs using the cluster I've configured from this aren't ready to deal with a pod security policy just yet, so I can understand if you don't want this in the final pull request.

**Comment by skinlayers** [comment]
@morgante The work around allowed me to create the infra, but I ran into the same problem when attempting to then run `make docker_test_integration `. I ended up making a few more changes to accommodate my immediate needs. I'll attempt to interactively step through the tests as described in the contribution guide.

File: autogen/safer-cluster/variables.tf.tmpl [hunk 4]
```diff
@@ -303,3 +303,9 @@ variable "enable_shielded_nodes" {
   default     = true
 }
 
+variable "enable_private_endpoint" {
+  type        = bool
+  description = "(Beta) Whether the master's internal IP address is used as the cluster endpoint"
+  default     = false
```
Comment (skinlayers) [inline autogen/safer-cluster/variables.tf.tmpl]: This has the opposite effect of what its name implies. In order to follow the GKE hardening guides recommendation at https://cloud.google.com/kubernetes-engine/docs/how-to/hardening-your-cluster#restrict_network_access_to_the_control_plane_and_nodes ("Public endpoint access enabled, master authorized networks enabled"), this `enable_private_endpoint` needs to be set to `false`.


See also https://stackoverflow.com/a/58651940/6720110

**Comment by skinlayers** [comment]
Security removed my GCP service account's org level permissions, so I'm stuck for now as I can't run the tests myself. I was hoping the fixes to the safer-cluster tests would resolve my issues, but no such luck.

**Comment by morgante** [comment]
@skinlayers FYI it's not required to have org level access to run the tests (that's just the easiest way). You can also run the tests in existing projects.

You can also populate a `test/source.sh` file with the appropriate TF variables. Ex. `TF_VAR_project_ids=['project-1', 'project-2']`.

You just need to populate a value for each output from the setup: https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/test/setup/outputs.tf

**Comment by morgante** [comment]
Alternatively, could we have @omazin take this over since he's already tweaking the autogen system?

**Comment by omazin** [comment]
@morgante I've accommodated the [devtools generate_modules helper](https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/pull/528) for the safer-cluster module generation. @skinlayers after merging https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/pull/528 and https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/389 you can safely remove the `helpers/generate_modules/generate_modules.py` file from the PR as it's going to be a part of the latest devtools image.

**Comment by skinlayers** [comment]
@morgante Ah! Thank you again for the testing information! And I am more than happy to let @omazin take over.

@omazin Any thoughts on the conversations I started further up?

**Comment by omazin** [comment]
@skinlayers, @morgante I think this can be merged, all tests passed successfully. Then after merging GoogleCloudPlatform/cloud-foundation-toolkit#528 the #389 can be merged since it has all required autogen configuration, including the `safer-cluster` module generation config.

**Comment by skinlayers** [comment]
I've pulled in the latest changes from master to fix a merge conflict in `autogen/main/README.md`.

**Pull request merged**